File: ruby/17_letter_combinations_of_a_phone_number.rb
# @param {String} digits
# @return {String[]}


# 根据电话拨号界面上每个数字下面的字母，给定字符串（2-9), 求每个数字对应的字母所有可能组合
# 使用map 方法， letters.map(&:chars) == letters.map { |str| str.chars }
# 所以首先根据m， 将digits 利用 .chars 变成数组后， 根据m对应表将2-9，转成对应的字符串（'abc')
# 接着 map letters 将 'abc' 转成对应的 .chars ['a', 'b', 'c']
# 然后 reduce， 使用product方法， 该方法组合两个数组内所有元素的情况
# 最后 使用数组的 join方法， 将['a', 'c'] = > 'ac'


def letter_combinations(digits)
    m = {
    '1' => '',     '2' => 'abc', '3' => 'def',
    '4' => 'ghi',  '5' => 'jkl', '6' => 'mno',
    '7' => 'pqrs', '8' => 'tuv', '9' => 'wxyz',
                   '0' => ''
    }

    return [] if digits.length == 0
    return m[digits[0]].chars if digits.length == 1

    letters = digits.chars.map { |digit| m[digit] }
    letters.map(&:chars).reduce(&:product).map(&:join)
end


nums = [1, 2, 3, 4]
letters = ['abc', 'def']

puts letters.map(&:chars).inspect
puts letters.map { |str| str.chars }.inspect
puts letters.map(&:chars).reduce(&:product).inspect
puts letters.map(&:chars).reduce(&:product).map(&:join).inspect


puts nums.join.class

File: cpp/17_letter_combinations_of_a_phone_number.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

/*
给定一个仅包含数字 2-9 的字符串，返回所有它能表示的字母组合。答案可以按 任意顺序 返回。

给出数字到字母的映射如下（与电话按键相同）。注意 1 不对应任何字母。
*/

// [ref](https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number/solution/hui-su-by-deeplcz-ymye/)
class Solution {
public:
	void helper(vector<string>& res, vector<string> strs, string& chosen, int index) {
		if (strs.size() == index) { 
			res.push_back(chosen);
			return;
		}
		string str = strs[index];
		for (int i = 0; i < str.size(); ++i) {
			chosen.push_back(str[i]);
			helper(res, strs, chosen, index+1);
			chosen.pop_back();
		}
	}

    vector<string> letterCombinations(string digits) {
        if (digits.empty()) { return {}; }
		unordered_map<int, string> um;
		um[2] = "abc";
		um[3] = "def";
		um[4] = "ghi";
		um[5] = "jkl";
		um[6] = "mno";
		um[7] = "pqrs";
		um[8] = "tuv";
		um[9] = "wxyz";
		
        vector<string> res;
		vector<string> strs;
		for (char ch : digits) {
			strs.push_back(um[ch - '0']);
		}
        string chosen;
		helper(res, strs, chosen, 0);

        return res;
    }
};
